packages/program/src/revert/revert-error.ts

```diff
@@ -0,0 +1,97 @@
+/* eslint-disable max-classes-per-file */
+import type { TransactionResultRevertReceipt } from '@fuel-ts/providers';
+import {
+  FAILED_ASSERT_EQ_SIGNAL,
+  FAILED_ASSERT_SIGNAL,
+  FAILED_REQUIRE_SIGNAL,
+  FAILED_SEND_MESSAGE_SIGNAL,
+  FAILED_TRANSFER_TO_ADDRESS_SIGNAL,
+  FAILED_UNKNOWN_SIGNAL,
+} from '@fuel-ts/transactions';
+
+export type RevertReason =
+  | 'RequireFailed'
+  | 'TransferToAddressFailed'
+  | 'SendMessageFailed'
+  | 'AssertEqFailed'
+  | 'AssertFailed'
+  | 'Unknown';
+
+const REVERT_MAP: { [signal: string]: RevertReason } = {
+  [FAILED_REQUIRE_SIGNAL]: 'RequireFailed',
+  [FAILED_TRANSFER_TO_ADDRESS_SIGNAL]: 'TransferToAddressFailed',
+  [FAILED_SEND_MESSAGE_SIGNAL]: 'SendMessageFailed',
+  [FAILED_ASSERT_EQ_SIGNAL]: 'AssertEqFailed',
+  [FAILED_ASSERT_SIGNAL]: 'AssertFailed',
+  [FAILED_UNKNOWN_SIGNAL]: 'Unknown',
+};
+
+const decodeRevertErrorCode = (
+  receipt: TransactionResultRevertReceipt
+): RevertReason | undefined => {
+  const signalHex = receipt.val.toHex();
+  return REVERT_MAP[signalHex] ? REVERT_MAP[signalHex] : undefined;
+};
+
+export class RevertError extends Error {
+  receipt: TransactionResultRevertReceipt;
+
+  constructor(receipt: TransactionResultRevertReceipt, reason: RevertReason) {
+    super(`The script reverted with reason ${reason}`);
+    this.name = 'RevertError';
+    this.receipt = receipt;
+  }
+
+  toString() {
+    const { id, ...r } = this.receipt;
+    return `${this.name}: ${this.message}
+    ${id}: ${JSON.stringify(r)}`;
+  }
+}
+
+export class RequireRevertError extends RevertError {
+  constructor(receipt: TransactionResultRevertReceipt, reason: RevertReason) {
+    super(receipt, reason);
+    this.name = 'RequireRevertError';
+  }
+}
+export class TransferToAddressRevertError extends RevertError {
+  constructor(receipt: TransactionResultRevertReceipt, reason: RevertReason) {
+    super(receipt, reason);
+    this.name = 'TransferToAddressRevertError';
+  }
+}
+export class SendMessageRevertError extends RevertError {
+  constructor(receipt: TransactionResultRevertReceipt, reason: RevertReason) {
+    super(receipt, reason);
+    this.name = 'SendMessageRevertError';
+  }
+}
+export class AssertFailedRevertError extends RevertError {
+  constructor(receipt: TransactionResultRevertReceipt, reason: RevertReason) {
+    super(receipt, reason);
+    this.name = 'AssertFailedRevertError';
+  }
+}
+
+export const revertErrorFactory = (
+  receipt: TransactionResultRevertReceipt
+): RevertError | undefined => {
+  const reason = decodeRevertErrorCode(receipt);
```

Comment:
I didn't understand by reading the code + tests, if this reason is going to include the log from the contract when using `require` as example, once the log is on the previous receipt and not on the val.

Reply:
When there is a required error, we could get the log and use it as the reason for removing the necessity of the developer to know that he needs to get the logs to see it. The idea is that the `log` for `require` is the reason why it had fail.